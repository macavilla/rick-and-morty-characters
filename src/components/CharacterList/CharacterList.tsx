import { CharacterType } from "@/types/character";

import CharacterCard from "../CharacterCard/CharacterCard";
export interface CharacterListProps {
  characters: CharacterType[];
  listId: 1 | 2;
  selectedCharacter: CharacterType | null;
  otherSelectedCharacter: CharacterType | null;
  onSelect: (c: CharacterType | null) => void;
}

const CharacterList = ({
  listId,
  characters,
  selectedCharacter,
  otherSelectedCharacter,
  onSelect,
}: CharacterListProps) => {
  return (
    <div className="">
      <h3 className="text-2xl">Character #{listId}</h3>

      <ul className="grid grid-cols-1 sm:grid-cols-2 gap-4 border-4">
        {characters.map((character: CharacterType) => (
          <CharacterCard
            key={character.id}
            character={character}
            isSelected={character === selectedCharacter}
            isDisabled={character === otherSelectedCharacter}
            onSelect={() =>
              character !== selectedCharacter
                ? onSelect(character)
                : onSelect(null)
            }
          />
        ))}
      </ul>
    </div>
  );
};

export default CharacterList;
