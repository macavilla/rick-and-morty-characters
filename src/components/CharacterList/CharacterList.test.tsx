import { screen, render, fireEvent } from "@testing-library/react";
import {
  mockCharacters,
  mockCharacterListProps,
} from "./__mockData__/CharacterList.mock";

import CharacterList from "./CharacterList";

describe("CharacterCard", () => {
  it("renders CharacterCard and matches snapshot", () => {
    const { container } = render(<CharacterList {...mockCharacterListProps} />);
    expect(container).toMatchSnapshot();
  });

  it("renders title and characters", () => {
    render(
      <CharacterList
        listId={1}
        characters={mockCharacters}
        selectedCharacter={null}
        otherSelectedCharacter={null}
        onSelect={jest.fn()}
      />
    );

    expect(screen.getByText("Character #1")).toBeInTheDocument();

    mockCharacters.forEach((character, index) => {
      expect(screen.getByText(mockCharacters[index].name)).toBeInTheDocument();
    });
  });

  it("calls onSelect when character is clicked", () => {
    const onSelectMock = jest.fn();

    render(
      <CharacterList
        listId={1}
        characters={mockCharacters}
        selectedCharacter={null}
        otherSelectedCharacter={null}
        onSelect={onSelectMock}
      />
    );

    fireEvent.click(screen.getByText(mockCharacters[0].name));
    expect(onSelectMock).toHaveBeenCalledWith(mockCharacters[0]);
  });

  it("calls onSelect(null) if character is already selected", () => {
    const onSelectMock = jest.fn();

    render(
      <CharacterList
        listId={1}
        characters={mockCharacters}
        selectedCharacter={mockCharacters[0]}
        otherSelectedCharacter={null}
        onSelect={onSelectMock}
      />
    );

    fireEvent.click(screen.getByText(mockCharacters[0].name));
    expect(onSelectMock).toHaveBeenCalledWith(null);
  });
});
