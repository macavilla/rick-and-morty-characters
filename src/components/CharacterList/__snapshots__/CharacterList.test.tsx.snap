// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`CharacterCard renders CharacterCard and matches snapshot 1`] = `
<div>
  <div
    class=""
  >
    <h3
      class="text-3xl text-center"
    >
      Character #
      1
    </h3>
    <ul
      class="grid grid-cols-1 sm:grid-cols-2 gap-2 border-3 rounded-4xl p-2 my-2 w-full"
    >
      <div
        class="
      p-2 border-2 rounded-3xl lg:flex lg:gap-3 
      cursor-not-allowed opacity-50
       ring-indigo-500 border-2 border-indigo-950 text-indigo-700  font-bold
    "
      >
        <div
          class="w-full lg:w-1/2"
        >
          <div
            class="relative"
          >
            <img
              alt="Rick Sanchez"
              class="w-full h-auto  rounded-2xl"
              data-nimg="1"
              decoding="async"
              height="300"
              loading="lazy"
              src="/_next/image?url=https%3A%2F%2Frickandmortyapi.com%2Fapi%2Fcharacter%2Favatar%2F1.jpeg&w=640&q=75"
              srcset="/_next/image?url=https%3A%2F%2Frickandmortyapi.com%2Fapi%2Fcharacter%2Favatar%2F1.jpeg&w=384&q=75 1x, /_next/image?url=https%3A%2F%2Frickandmortyapi.com%2Fapi%2Fcharacter%2Favatar%2F1.jpeg&w=640&q=75 2x"
              style="color: transparent;"
              width="300"
            />
            <div
              class="flex items-center gap-2 absolute top-1 left-1 border-1 bg-foreground border-background text-background rounded-full py-0.5 px-1.5"
            >
              <div
                aria-hidden="true"
                class="rounded-full size-3  bg-green-500"
              />
              <p
                class="capitalize text-xs"
              >
                Alive
              </p>
            </div>
          </div>
        </div>
        <div
          class="lg:w-1/2 py-2  mt-2 lg:m-0 lg:p-0  lg:pl-3"
        >
          <h4
            class="text-xl lg:text-2xl mb-2"
          >
            Rick Sanchez
          </h4>
          <p
            class="italic text-lg"
          >
            Human
          </p>
        </div>
      </div>
    </ul>
  </div>
</div>
`;
